homeassistant/components/restore_state.py

```diff
@@ -0,0 +1,86 @@
+"""
+Support for restoring entity states on startup.
+
+Component that records all events and state changes. Allows other components
+to query this database.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/recorder/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, split_entity_id
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START, ATTR_RESTORED_STATE)
+from homeassistant.components.history import (
+    CONF_DOMAINS, CONF_EXCLUDE, CONF_ENTITIES, CONF_INCLUDE,
+    Filters, get_states, last_recorder_run)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import ConfigType
+import homeassistant.util.dt as dt_util
+
+DOMAIN = 'restore_state'
+
+DEPENDENCES = ['recorder']
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: cv.FILTER_SCHEMA,
+}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+_ALWAYS_OVERWRITE = (
```

Comment:
Ever since I've stumbled upon this PR I've been wondering how the restore_state component is going to push the state to individual entities, as components usually don't have any methods like set_state (at least there's no common API for that in Entity class) and internal implementation of keeping state can be different in each component or platform implementation.

I like the idea of a new hook. I don't like proposed implementation of calling `self.restore_state` from the callback/hook. That way individual platforms would be able to only restore known attributes or overwrite the hook itself.

Maybe it would be better to either pass whole `state.attributes` as a single argument or pass it as `**state.attributes`. That way a component doesn't need to assume which attributes should be passed to individual platforms.